src/peer_manager.rs

```diff
@@ -505,21 +514,43 @@ impl PeerManager {
             return Err(RoutingError::UnknownCandidate);
         }
 
-        Ok((*new_pub_id, *new_client_auth))
+        Ok((
+            *new_pub_id,
+            OnlinePayload {
+                client_auth: *new_client_auth,
+                old_public_id: *old_pub_id,
+            },
+        ))
+    }
+
+    /// Handles accumulated candidate approval. Marks the candidate as `Approved`.
+    /// If the candidate was already purged or is unexpected, return false.
+    pub fn handle_candidate_online_event(
+        &mut self,
+        new_pub_id: &PublicId,
+        old_pub_id: &PublicId,
+    ) -> bool {
+        if self.candidate.old_pub_id() == Some(old_pub_id) {
+            // Known candidate was accepted.
+            // Ignore any information that could be different stored in candidate.
+            // Do not accept candidate until we complete our current one.
+            self.candidate = Candidate::ApprovedWaitingSectionInfo {
+                new_pub_id: *new_pub_id,
+            };
+            true
+        } else {
+            // Unkwown candidate, Candidate was purged before: refuse it.
+            false
+        }
     }
 
-    /// Handles accumulated candidate approval. Marks the candidate as `Approved` and returns if the
-    /// candidate is connected or `Err` if the peer is not the candidate or we're missing its info.
+    /// Handles approved accumulated candidate. Returns if the candidate is connected or
+    /// `Err` if the peer is not the candidate or we're missing its info.
     pub fn handle_candidate_approval(
```

Comment:
So this function now seems to just return whether the peer is connected. I'd consider renaming it to `is_peer_connected` or something like that.